### PR TITLE
GPII-3136: Chart and GCP module for k8s-snapshots

### PR DESCRIPTION
**Pull request by natarajaya**
While this does the job perfectly (creates encrypted snapshots from Kubernetes-managed disks with proper annotation), restoration process is not 100% clear to me at the moment. I'll investigate it further and hopefully come come up with detailed backup restore plan.

### CODE REVIEW

File: gcp/live/dev/k8s/kube-system/k8s-snapshots/terraform.tfvars [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# ↓ Module metadata
+terragrunt = {
+  terraform {
+    source = "/project/modules//k8s-snapshots"
+  }
+
+  dependencies {
```
Comment (mrtyler) [inline gcp/live/dev/k8s/kube-system/k8s-snapshots/terraform.tfvars]: This dependency is not needed.

File: gcp/modules/k8s-snapshots/main.tf [hunk 2]
```diff
@@ -0,0 +1,45 @@
+terraform {
+  backend "gcs" {}
+}
+
+variable "prd_deltas" {
```
Comment (mrtyler) [inline gcp/modules/k8s-snapshots/main.tf]: Per-environment settings should be set in `live/dev`, `live/stg`, etc.

...but I don't think these settings belong in this module. See below.

File: gcp/modules/k8s-snapshots/main.tf [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+  chart_name = "../../../../../charts/k8s-snapshots"
+}
+
+resource "null_resource" "enable_backups" {
+  depends_on = ["module.k8s-snapshots"]
+
+  provisioner "local-exec" {
+    command = <<EOF
+      for PV in $(kubectl get pv -o jsonpath="{range .items[*]}{.metadata.name}{' '}"); do
```
Comment (mrtyler) [inline gcp/modules/k8s-snapshots/main.tf]: Can these annotations be set by the PVs themselves, as they are on the AWS side? The approach here seems backwards to me.

File: common/charts/k8s-snapshots/values.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# Default values for k8s-snapshots.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+replicaCount: 1
+
+image:
+  repository: elsdoerfer/k8s-snapshots
```
Comment (mrtyler) [inline common/charts/k8s-snapshots/values.yaml]: If the goal is to use this chart for the AWS side, note that upstream is broken and we are using an image I built: https://github.com/gpii-ops/gpii-infra/blob/master/aws/modules/deploy/05-k8s-snapshots-deploy.yml#L15

File: gcp/live/dev/k8s/kube-system/k8s-snapshots/terraform.tfvars (hunk 1 above)
Comment (natarajaya) [reply gcp/live/dev/k8s/kube-system/k8s-snapshots/terraform.tfvars]: It was needed, because we wanted to patch volumes after they are created. Not anymore.

File: gcp/modules/k8s-snapshots/main.tf (hunk 2 above)
Comment (natarajaya) [reply gcp/modules/k8s-snapshots/main.tf]: Removed.

File: gcp/modules/k8s-snapshots/main.tf (hunk 3 above)
Comment (natarajaya) [reply gcp/modules/k8s-snapshots/main.tf]: Removed after switching from K8s-managed vols in https://github.com/gpii-ops/gpii-infra/pull/51

File: common/charts/k8s-snapshots/values.yaml (hunk 4 above)
Comment (natarajaya) [reply common/charts/k8s-snapshots/values.yaml]: I just checked and it seems that your fork is working properly for GCP. However, I want to keep upstream version in the chart itself. We can still use it for AWS, image attributes are configurable via custom values.

**Comment by natarajaya** [comment]
@mrtyler @amatas Please check out updated version!

File: common/charts/k8s-snapshots/values.yaml (hunk 4 above)
Comment (mrtyler) [reply common/charts/k8s-snapshots/values.yaml]: My "fork" only reverts a commit upgrading boto, which didn't work. See the comment in the linked code for details.

I agree with keeping the official upstream image in the chart. I'm only warning you that the official upstream image won't work for AWS, so when you switch AWS over to the chart version you'll need to deal with that.

File: gcp/modules/k8s-snapshots/values.yaml [hunk 5]
```diff
@@ -0,0 +1 @@
+useClaimName: false
```
Comment (mrtyler) [inline gcp/modules/k8s-snapshots/values.yaml]: Why is this needed? We use `true` on the AWS side, which produces much better Snapshot names.

**Comment by mrtyler** [comment]
Updates look good. One remaining question about `useClaimName`.

File: common/charts/k8s-snapshots/values.yaml (hunk 4 above)
Comment (natarajaya) [reply common/charts/k8s-snapshots/values.yaml]: Alright, I am going to apply fixes to AWS part in this PR as well.

File: gcp/modules/k8s-snapshots/values.yaml (hunk 5 above)
Comment (natarajaya) [reply gcp/modules/k8s-snapshots/values.yaml]: Related to K8s-managed vols vs Terraform-managed.

**Comment by natarajaya** [comment]
@mrtyler I updated AWS part to use common chart.

**Comment by mrtyler** [comment]
LGTM!
